File: go/okapi/Oberon.go
package okapi

import okapi "github.com/trinsic-id/okapi/go/okapi/proto"

type Oberoner interface {
	CreateKey(request *okapi.CreateOberonKeyRequest) (*okapi.CreateOberonKeyResponse, error)
	CreateToken(request *okapi.CreateOberonTokenRequest) (*okapi.CreateOberonTokenResponse, error)
	BlindToken(request *okapi.BlindOberonTokenRequest) (*okapi.BlindOberonTokenResponse, error)
	UnblindToken(request *okapi.UnBlindOberonTokenRequest) (*okapi.UnBlindOberonTokenResponse, error)
	CreateProof(request *okapi.CreateOberonProofRequest) (*okapi.CreateOberonProofResponse, error)
	VerifyProof(request *okapi.VerifyOberonProofRequest) (*okapi.VerifyOberonProofResponse, error)
}

func Oberon() Oberoner {
	return &oberon{}
}

type oberon struct{}

func (d *oberon) CreateKey(request *okapi.CreateOberonKeyRequest) (*okapi.CreateOberonKeyResponse, error) {
	response := okapi.CreateOberonKeyResponse{}
	err := callOkapiNative(request, &response, oberonCreateKey)
	return &response, err
}

func (d *oberon) CreateToken(request *okapi.CreateOberonTokenRequest) (*okapi.CreateOberonTokenResponse, error) {
	response := okapi.CreateOberonTokenResponse{}
	err := callOkapiNative(request, &response, oberonCreateToken)
	return &response, err
}

func (d *oberon) BlindToken(request *okapi.BlindOberonTokenRequest) (*okapi.BlindOberonTokenResponse, error) {
	response := okapi.BlindOberonTokenResponse{}
	err := callOkapiNative(request, &response, oberonBlindToken)
	return &response, err
}

func (d *oberon) UnblindToken(request *okapi.UnBlindOberonTokenRequest) (*okapi.UnBlindOberonTokenResponse, error) {
	response := okapi.UnBlindOberonTokenResponse{}
	err := callOkapiNative(request, &response, oberonUnBlindToken)
	return &response, err
}

func (d *oberon) CreateProof(request *okapi.CreateOberonProofRequest) (*okapi.CreateOberonProofResponse, error) {
	response := okapi.CreateOberonProofResponse{}
	err := callOkapiNative(request, &response, oberonCreateProof)
	return &response, err
}

func (d *oberon) VerifyProof(request *okapi.VerifyOberonProofRequest) (*okapi.VerifyOberonProofResponse, error) {
	response := okapi.VerifyOberonProofResponse{}
	err := callOkapiNative(request, &response, oberonVerifyProof)
	return &response, err
}

File: ruby/lib/okapi/oberon.rb
# frozen_string_literal: true

module Okapi
  # Oberon module
  module Oberon
    def self.create_key(request)
      Okapi.verify_type(request, Okapi::Security::V1::CreateOberonKeyRequest)
      Okapi.ffi_call('oberon_create_key', request, Okapi::Security::V1::CreateOberonKeyResponse)
    end

    def self.create_token(request)
      Okapi.verify_type(request, Okapi::Security::V1::CreateOberonTokenRequest)
      Okapi.ffi_call('oberon_create_token', request, Okapi::Security::V1::CreateOberonTokenResponse)
    end

    def self.blind_token(request)
      Okapi.verify_type(request, Okapi::Security::V1::BlindOberonTokenRequest)
      Okapi.ffi_call('oberon_blind_token', request, Okapi::Security::V1::BlindOberonTokenResponse)
    end

    def self.unblind_token(request)
      Okapi.verify_type(request, Okapi::Security::V1::UnBlindOberonTokenRequest)
      Okapi.ffi_call('oberon_unblind_token', request, Okapi::Security::V1::UnBlindOberonTokenResponse)
    end

    def self.verify_token(request)
      Okapi.verify_type(request, Okapi::Security::V1::VerifyOberonTokenRequest)
      Okapi.ffi_call('oberon_verify_token', request, Okapi::Security::V1::VerifyOberonTokenResponse)
    end

    def self.create_proof(request)
      Okapi.verify_type(request, Okapi::Security::V1::CreateOberonProofRequest)
      Okapi.ffi_call('oberon_create_proof', request, Okapi::Security::V1::CreateOberonProofResponse)
    end

    def self.verify_proof(request)
      Okapi.verify_type(request, Okapi::Security::V1::VerifyOberonProofRequest)
      Okapi.ffi_call('oberon_verify_proof', request, Okapi::Security::V1::VerifyOberonProofResponse)
    end
  end
end